lxd/networks_utils.go

```diff
@@ -1170,73 +1172,114 @@ func networkGetMacSlice(hwaddr string) []string {
 	return buf
 }
 
-func networkClearLease(s *state.State, name string, network string, hwaddr string) error {
+func networkClearLease(s *state.State, name string, network string, hwaddr string, timeNow time.Time) error {
 	leaseFile := shared.VarPath("networks", network, "dnsmasq.leases")
 
 	// Check that we are in fact running a dnsmasq for the network
 	if !shared.PathExists(leaseFile) {
 		return nil
 	}
 
-	// Restart the network when we're done here
-	n, err := networkLoadByName(s, network)
+	// Convert MAC string to bytes to avoid any case comparison issues later.
+	srcMAC, err := net.ParseMAC(hwaddr)
 	if err != nil {
 		return err
 	}
-	defer func() {
-		err := n.Start()
-		if err != nil {
-			logger.Errorf("Failed to reload network '%s': %v", network, err)
-		}
-	}()
 
-	// Stop dnsmasq
-	err = networkKillDnsmasq(network, false)
+	iface, err := net.InterfaceByName(network)
 	if err != nil {
 		return err
 	}
 
-	// Mangle the lease file
-	leases, err := ioutil.ReadFile(leaseFile)
+	// Get IPv4 and IPv6 address of interface running dnsmasq on host.
+	addrs, err := iface.Addrs()
 	if err != nil {
 		return err
 	}
 
-	fd, err := os.Create(leaseFile)
+	var dstIPv4, dstIPv6 net.IP
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			return err
+		}
+		if !ip.IsGlobalUnicast() {
+			continue
+		}
+		if ip.To4() == nil {
+			dstIPv6 = ip
+		} else {
+			dstIPv4 = ip
+		}
+	}
+
+	// Ensure we have waited at least 3 seconds for dnsmasq leases file to be
+	// written out before scanning it to detect leases that need releasing.
+	// Otherwise we will likely leave orphaned leases in the file.
+	// But don't sleep if more than that has already passed.
+	if time.Since(timeNow) < (time.Second * 3) {
```

Comment:
@stgraber I am not entirely happy with this solution. I have made sure to only delay Delete operations on containers that have bridged devices using a local dnsmasq, and it will only delay it once, not once per device.

However the delay picked, 3s, is rather arbitrary based on my own tests to avoid the race between a container starting up, issuing a DHCP SOLICIT packet just before being stopped by Stop() before Delete() is called.

By the time Delete() is called we cannot not know whether a container had an IP allocated at all, so we cannot know whether there is a DHCP that needs to be released. 

This approach just waits and *hopes* that dnsmasq has flushed the leases file by the time we parse it to check for any leases that need releasing. Most of the time this is fine, but when starting/deleting containers in quick success, without this sleep, it can cause orphaned leases.

Reply:
@stgraber I found the issue, it was caused by dnsmasq trying to ping the IPv4 address it was considering offering to the container, which delayed the allocation and queued up the DHCPv6 requests behind it for several seconds. Once the IP was detected as not in use, it was offered to the container, but by this time it was shut down. Then the IPv6 SOLICIT packets were processed and the DHCPv6 lease was created, but after the container had been deleted. 

Adding the `--no-ping` flag to dnsmasq fixes this issue.